pkg/storage/engine.go

```diff
@@ -961,6 +961,15 @@ type Batch interface {
 	// engine. This is a noop unless the batch was created via NewBatch(). If
 	// sync is true, the batch is synchronously committed to disk.
 	Commit(sync bool) error
+	// CommitNoSyncWait atomically applies any batched updates to the underlying
+	// engine and initiates a synchronous disk write, but does not wait for that
+	// write to complete. The caller must call SyncWait to wait for the fsync to
+	// complete. The caller must not Close the Batch without first calling
+	// SyncWait.
+	CommitNoSyncWait() error
+	// SyncWait waits for the disk write initiated by a call to CommitNoSyncWait
+	// to complete.
```

Comment:
When this completes it provides a bit more, right? It also means all previous writes can be considered completed.

Reply:
Also maybe something about what happens on error here and in CommitNoSyncWait - I assume it means that fsyncgate-style all bets are off and it'd be better to crash.

Reply:
> When this completes it provides a bit more, right? It also means all previous writes can be considered completed.

In practice, that is currently true, but I don't think we'd want to make such a promise. `pebble.DB.ApplyNoSyncWait` avoids documenting the guarantees that one sync-ed write makes regarding the durability of other prior, unsynced writes. I'd imagine that such guarantees could be violated if we ever partitioned the WAL or permitted other forms of reordering within the pebble commit pipeline.

> Also maybe something about what happens on error here and in CommitNoSyncWait

Right now, we never actually return an error. I've added a trio of TODOs to address this after the discussion in https://cockroachlabs.slack.com/archives/CAC6K3SLU/p1674678349690899.